tfhe/src/keycache/mod.rs

```diff
@@ -186,32 +200,44 @@ pub mod utils {
                         .map(|param_key| param_key.1.clone())
                 };
 
-                if let Some(shared_cell) = maybe_shared_cell {
-                    shared_cell.inner.get_or_init(load_from_persistent_storage);
-                    Ok(shared_cell)
-                } else {
-                    Err(())
-                }
+                maybe_shared_cell.map_or_else(
+                    || {
+                        let key_not_found = {
+                            self.memory_storage
+                                .read()
+                                .unwrap()
+                                .iter()
+                                .all(|(p, _)| *p != param)
+                        };
+
+                        if key_not_found {
+                            let shared_key = SharedKey {
+                                inner: Arc::new(OnceLock::new()),
+                            };
+                            shared_key.inner.get_or_init(load_from_persistent_storage);
+                            {
+                                // we only hold a write lock for a short duration to push the lazily
+                                // evaluated key without actually evaluating the key
+                                let mut memory_storage = self.memory_storage.write().unwrap();
+                                if memory_storage.iter().all(|(p, _)| *p != param) {
+                                    memory_storage.push((param, shared_key));
```

Comment:
here we can clone the shared key (it is an Arc so small in memory) instead of reloading it below